homeassistant/components/opentherm_gw/binary_sensor.py

```diff
@@ -3,116 +3,51 @@
-            DEVICE_CLASS_PROBLEM, "Boiler Water Overtemperature"],
-        gw_vars.DATA_REMOTE_TRANSFER_DHW: [
-            None, "Remote Hot Water Setpoint Transfer Support"],
-        gw_vars.DATA_REMOTE_TRANSFER_MAX_CH: [
-            None, "Remote Maximum Central Heating Setpoint Write Support"],
-        gw_vars.DATA_REMOTE_RW_DHW: [
-            None, "Remote Hot Water Setpoint Write Support"],
-        gw_vars.DATA_REMOTE_RW_MAX_CH: [
-            None, "Remote Central Heating Setpoint Write Support"],
-        gw_vars.DATA_ROVRD_MAN_PRIO: [
-            None, "Remote Override Manual Change Priority"],
-        gw_vars.DATA_ROVRD_AUTO_PRIO: [
-            None, "Remote Override Program Change Priority"],
-        gw_vars.OTGW_GPIO_A_STATE: [None, "Gateway GPIO A State"],
-        gw_vars.OTGW_GPIO_B_STATE: [None, "Gateway GPIO B State"],
-        gw_vars.OTGW_IGNORE_TRANSITIONS: [None, "Gateway Ignore Transitions"],
-        gw_vars.OTGW_OVRD_HB: [None, "Gateway Override High Byte"],
-    }
+    gw_dev = hass.data[DATA_OPENTHERM_GW][DATA_GATEWAYS][discovery_info]
     sensors = []
-    for var in discovery_info:
-        device_class = sensor_info[var][0]
-        friendly_name = sensor_info[var][1]
-        entity_id = async_generate_entity_id(ENTITY_ID_FORMAT, var, hass=hass)
-        sensors.append(OpenThermBinarySensor(entity_id, var, device_class,
-                                             friendly_name))
+    for var in gw_dev.binary_sensors:
+        device_class = BINARY_SENSOR_INFO[var][0]
+        friendly_name_format = BINARY_SENSOR_INFO[var][1]
+        sensors.append(OpenThermBinarySensor(gw_dev, var, device_class,
+                                             friendly_name_format))
     async_add_entities(sensors)
 
 
 class OpenThermBinarySensor(BinarySensorDevice):
     """Represent an OpenTherm Gateway binary sensor."""
 
-    def __init__(self, entity_id, var, device_class, friendly_name):
+    def __init__(self, gw_dev, var, device_class, friendly_name_format):
         """Initialize the binary sensor."""
-        self.entity_id = entity_id
+        self.entity_id = async_generate_entity_id(
+            ENTITY_ID_FORMAT, '{}_{}'.format(var, gw_dev.gw_id),
+            hass=gw_dev.hass)
+        self._gateway = gw_dev
         self._var = var
         self._state = None
         self._device_class = device_class
-        self._friendly_name = friendly_name
+        self._friendly_name = friendly_name_format.format(gw_dev.name)
 
     async def async_added_to_hass(self):
         """Subscribe to updates from the component."""
         _LOGGER.debug(
             "Added OpenTherm Gateway binary sensor %s", self._friendly_name)
-        async_dispatcher_connect(self.hass, SIGNAL_OPENTHERM_GW_UPDATE,
+        async_dispatcher_connect(self.hass, self._gateway.update_signal,
                                  self.receive_report)
 
-    async def receive_report(self, status):
+    def receive_report(self, status):
```

Comment:
This must be decorated with `@callback` imported from core.py to make sure this callback is run in the event loop.

Reply:
Fixed in 83ea38d, and applied in the other modules as well (sensor.py, climate.py).